File mismatch investigation

**Impact of the bug**
Identify issue with stuck PnR workflows and allow them to progress.

**Describe the bug**
As was described in [PnR#41](https://gitlab.cern.ch/CMSToolsIntegration/operations/-/issues/41) we have certain level of workflows which have file mismatch between DBS and Rucio and they never get announced for months. The filemismatch usually happens when there are output files in DBS but not recorded by rucio, here is one example:
```
file: /store/data/Run2022D/Muon/RAW-RECO/ZMu-10Dec2022-v1/2550000/025d55a6-8e48-4695-9087-80bf9681229b.root
```
for WF:  `pdmvserv_Run2022D_Muon_10Dec2022_221209_230856_3545`. A recent list of workflows having such issue can be found [here](https://dmytro.web.cern.ch/dmytro/cmsprodmon/workflows.php?prep_id=task_EGM-Run3Winter23wmLHEGS-00001).

**How to reproduce it**
Most likely we need to perform debugging of agent logs and its local databases to identify the issue with stuck workflows.

**Expected behavior**
We should have a file match between DBS and Rucio to allow WF to progress.

**Additional context and error message**

Alan gave me some pointers and I would like to update on my findings:
- I start with the following LFN: `/store/mc/RunIISummer20UL18RECO/G1Jet_LHEGpT-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/AODSIM/106X_upgrade2018_realistic_v11_L1v1-v2/2820004/1D29A34F-EB43-D94D-840F-845115A852A6.root`
- it is defined in this block `/G1Jet_LHEGpT-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/RunIISummer20UL18RECO-106X_upgrade2018_realistic_v11_L1v1-v2/AODSIM#c56456c6-2aac-41e3-b7fc-aa3a569ac531`
- it belongs to this workflow: `cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766`
- which runs on these agents: `cmsgwms-submit3.fnal.gov,cmsgwms-submit4.fnal.gov,cmsgwms-submit5.fnal.gov,cmsgwms-submit6.fnal.gov,cmsgwms-submit7.fnal.gov,cmsgwms-submit8.fnal.gov,vocms0254.cern.ch,vocms0255.cern.ch,vocms0256.cern.ch,vocms0281.cern.ch,vocms0282.cern.ch,vocms0283.cern.ch`
- where using single agent `cmsgwms-submit5` and its log area `data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir`
- we can find this workflow over here `/data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766`
- and scanning for its logs using the following script:
```
[cmsdataops@cmsgwms-submit5 current]$ cat find_entries.sh
#!/bin/bash
idir=$1
pat=$2
files=`find $1 -name Job*.tar.bz2`
for fname in $files;
do
    results=`bzgrep "$pat" $fname`
    if [ -n "$results" ]; then
        echo "Pattern \"$pat\" found in $fname"
    fi
done
```
- we can see that pattern `stage out failed` can be found in the following logs:
```
[cmsdataops@cmsgwms-submit5 current]$ ./find_entries.sh $idir "$pat"
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711941.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711681.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711751.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711993.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711515.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711466.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711422.tar.bz2
Pattern "stage out failed" found in /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711292.tar.bz2
```
Now, if we look into individual log, e.g.
```
bzcat /data/srv/wmagent/current/install/wmagentpy3/JobArchiver/logDir/c/cmsunified_task_SMP-RunIISummer20UL18wmLHEGEN-00542__v1_T_230314_093054_8766/JobCluster_3711/Job_3711422.tar.bz2 | less
```
we can see the following failures:
```
2023-04-19 05:43:42,959:ERROR:StageOutImpl:Attempt 0 to stage out failed.
Automatically retrying in 300 secs
 Error details:
<@========== WMException Start ==========@>
Exception Class: StageOutError
Message: Command exited non-zero, ExitCode:151
Output: stdout: Local File Size is: 8393858
Remote File Size is: 0
Local File Checksum is: c3e0a807
Remote File Checksum is:
ERROR: Size or Checksum Mismatch between local and SE

stderr: Run: [ERROR] Operation expired:  (destination)
[ERROR] Server responded with an error: [3013] No checksum available for this file.


        ClassName : None
        ModuleName : WMCore.Storage.StageOutError
        MethodName : __init__
        ClassInstance : None
        FileName : /srv/job/WMCore.zip/WMCore/Storage/StageOutError.py
        LineNumber : 32
        ErrorNr : 0
        Command : LOCAL_SIZE=`stat -c%s "/srv/job/WMTaskSpace/cmsRun5/AODSIMoutput.root"`
...
Command exited non-zero, ExitCode:151
Output: stdout: Local File Size is: 8393858
Remote File Size is: 0
Local File Checksum is: c3e0a807
Remote File Checksum is:
ERROR: Size or Checksum Mismatch between local and SE

stderr: Run: [ERROR] Operation expired:  (destination)
[ERROR] Server responded with an error: [3013] No checksum available for this file.


ERROR:root:Attempt 0 to stage out failed.
...
echo "Local File Size is: $LOCAL_SIZE"
xrdcp --force --nopbar -p --cksum adler32:c3e0a807  "/srv/job/WMTaskSpace/cmsRun5/AODSIMoutput.root"  "root://cmsdcadisk.fnal.gov//dcache/uscmsdisk/store/unmerged/RunIISummer20UL18RECO/G1Jet_LHEGpT-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/AODSIM/106X_upgrade2018_realistic_v11_L1v1-v2/50568/172D2C31-CA56-8949-8FD4-50BCA1D033F5.root"
REMOTE_SIZE=`xrdfs 'cmsdcadisk.fnal.gov' stat '/dcache/uscmsdisk/store/unmerged/RunIISummer20UL18RECO/G1Jet_LHEGpT-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/AODSIM/106X_upgrade2018_realistic_v11_L1v1-v2/50568/172D2C31-CA56-8949-8FD4-50BCA1D033F5.root' | grep Size | sed -r 's/.*Size:[ ]*([0-9]+).*/\1/'`
echo "Remote File Size is: $REMOTE_SIZE"
echo "Local File Checksum is: c3e0a807"
REMOTE_XS=`xrdfs 'cmsdcadisk.fnal.gov' query checksum '/dcache/uscmsdisk/store/unmerged/RunIISummer20UL18RECO/G1Jet_LHEGpT
-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/AODSIM/106X_upgrade2018_realistic_v11_L1v1-v2/50568/172D2C3
1-CA56-8949-8FD4-50BCA1D033F5.root' | grep -i adler32 | sed -r 's/.*[adler|ADLER]32[ ]*([0-9a-fA-F]{8}).*/\1/'`
echo "Remote File Checksum is: $REMOTE_XS"
if [ $REMOTE_SIZE ] && [ $REMOTE_XS ] && [ $LOCAL_SIZE == $REMOTE_SIZE ] && [ 'c3e0a807' == $REMOTE_XS ]; then exit 0; els
e echo "ERROR: Size or Checksum Mismatch between local and SE"; xrdfs 'cmsdcadisk.fnal.gov' rm /dcache/uscmsdisk/store/unm
erged/RunIISummer20UL18RECO/G1Jet_LHEGpT-50To100_PhotonInBarrel_Mjj-300_TuneCP5_13TeV-amcatnlo-pythia8/AODSIM/106X_upgrade
2018_realistic_v11_L1v1-v2/50568/172D2C31-CA56-8949-8FD4-50BCA1D033F5.root ; exit 60311 ; fi
        ExitCode : 151
        ErrorCode : 60311
        ErrorType : GeneralStageOutFailure
```

Since we have different agents involved I need more information what we should spot in a log files. The more precise regexp you can provide the better, easier and faster we can find this needle.  May be we should look at specific code or message. Please provide any valuable input you can give or know.

--- a/src/python/WMCore/Storage/Backends/GFAL2Impl.py
+++ b/src/python/WMCore/Storage/Backends/GFAL2Impl.py
@@ -27,6 +27,8 @@ class GFAL2Impl(StageOutImpl):
         self.setups = "env -i X509_USER_PROXY=$X509_USER_PROXY JOBSTARTDIR=$JOBSTARTDIR bash -c '%s'"
         self.removeCommand = self.setups % '. $JOBSTARTDIR/startup_environment.sh; date; gfal-rm -t 600 %s '
         self.copyCommand = self.setups % '. $JOBSTARTDIR/startup_environment.sh; date; gfal-copy -t 2400 -T 2400 -p %(checksum)s %(options)s %(source)s %(destination)s'
+        self.copyOpts = '-t 2400 -T 2400 -p -v --abort-on-failure %(checksum)s %(options)s %(source)s %(destination)s'
+        self.copyCommand = self.setups % ('. $JOBSTARTDIR/startup_environment.sh; date; gfal-copy ' + self.copyOpts)
 
     def createFinalPFN(self, pfn):
         """
@@ -87,6 +89,8 @@ class GFAL2Impl(StageOutImpl):
           -T   global timeout for the transfer operation
           -p   if the destination directory does not exist, create it
           -K   checksum algorithm to use, or algorithm:value
+          -v   enable the verbose mode (-v for warning level)
+          --abort-on-failure  abort the whole copy as soon as one failure is encountered
         """
         result = "#!/bin/bash\n"
 

--- a/test/python/WMCore_t/Storage_t/Backends_t/GFAL2Impl_t.py
+++ b/test/python/WMCore_t/Storage_t/Backends_t/GFAL2Impl_t.py
@@ -17,8 +17,8 @@ class GFAL2ImplTest(unittest.TestCase):
         removeCommand = "env -i X509_USER_PROXY=$X509_USER_PROXY JOBSTARTDIR=$JOBSTARTDIR bash -c " \
                         "'. $JOBSTARTDIR/startup_environment.sh; date; gfal-rm -t 600 %s '"
         copyCommand = "env -i X509_USER_PROXY=$X509_USER_PROXY JOBSTARTDIR=$JOBSTARTDIR bash -c '" \
-                      ". $JOBSTARTDIR/startup_environment.sh; date; gfal-copy -t 2400 -T 2400 " \
-                      "-p %(checksum)s %(options)s %(source)s %(destination)s'"
+                      ". $JOBSTARTDIR/startup_environment.sh; date; gfal-copy -t 2400 -T 2400 -p " \
+                      "-v --abort-on-failure %(checksum)s %(options)s %(source)s %(destination)s'"
         self.assertEqual(removeCommand, testGFAL2Impl.removeCommand)
         self.assertEqual(copyCommand, testGFAL2Impl.copyCommand)
 

